src/controller/CHIPDeviceControllerFactory.h

```diff
@@ -138,6 +139,41 @@ class DeviceControllerFactory
     //
     void ReleaseSystemState() { mSystemState->Release(); }
 
+    class ControllerFabricDelegate final : public FabricTableDelegate
+    {
+    public:
+        ControllerFabricDelegate() {}
+        ControllerFabricDelegate(SessionManager * sessionManager) : FabricTableDelegate(true), mSessionManager(sessionManager) {}
+
+        CHIP_ERROR Init(SessionManager * sessionManager)
+        {
+            VerifyOrReturnError(sessionManager != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+            mSessionManager = sessionManager;
+            return CHIP_NO_ERROR;
+        };
+
+        void OnFabricDeletedFromStorage(CompressedFabricId compressedId, FabricIndex fabricIndex) override
+        {
+            if (mSessionManager != nullptr)
```

Comment:
I wonder whether we could share more of this code with Server somehow....  Followup for that, perhaps?  Having this cleanup code in multiple places makes it too likely for them to get out of sync.

Reply:
CHIPDeviceControllerFactory and Server .cpp could use a lot of clean up and sync since there is a lot of duplications inside. But yes this should be done in a follow up PR